### ISSUE TITLE
Impersonate-User / Impersonate-Group / Impersonate-Extra-* are not configurable for DefaultOpenShiftClient (works fine with DefaultKubernetesClient)

### ISSUE THREAD

**Issue by ibuziuk** [fabric8io/kubernetes-client#1085]
`Impersonate-User / Impersonate-Group / Impersonate-Extra-` are meant to be supported by both OpenShift & Kubernetes clients via `ImpersonatorInterceptor` [1]. However, the following sample with setting `Impersonate-User` header works only for Kubernetes client:

```
       ConfigBuilder configBuilder =
                  new ConfigBuilder()
                      .withMasterUrl(masterUrl)
                      .withNamespace("user-namespace")
                      .withTrustCerts(true)
                      .withOauthToken(cheServiceAccountToken);
                   

          Config config = configBuilder.build();
          config.getRequestConfig().setImpersonateUsername("username");

          KubernetesClient client = new DefaultKubernetesClient(config);
          List<Pod> pods = client.pods().list().getItems();
          for (Pod pod : pods) {
              System.out.println(pod.getMetadata().getName());
          }
```
If  `KubernetesClient client = new DefaultKubernetesClient(config);` from the sample would be changed to `DefaultOpenShiftClient` the Impersonate header would not be set by `ImpersonatorInterceptor`[1] during request processing. The reason why  `Impersonate-*` headers are not set for OpenShiftClient is because `Config` [2] constructor that is used internally during transformation to `OpenShiftConfig` does not honour `Impersonate` settings during building `RequestConfig` [3].  RequestConfig is later used in `ImpersonatorInterceptor` for setting headers.

[1] https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/ImpersonatorInterceptor.java#L36
[2] https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Config.java#L213
[3] https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/RequestConfig.java#L56

**Comment by ibuziuk** [comment]
@lulf hi, could you please comment on this issue since you seem to be most aware of `ImpersonatorInterceptor` details? This issue currently blocks lockdown of che namespaces on openshift.io  - https://github.com/redhat-developer/rh-che/pull/688 The fix seems to be pretty straight-forward (adding support of Impersonate settings in the `RequestConfig` [1] constructor). However, this fix would probably be treated as API change that is not suitable for service releases AFAIK

[1] https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/RequestConfig.java#L56

**Comment by lulf** [comment]
@ibuziuk IIRC, the impersonation support was limited to a per-request setting which is why it wasn't added to the RequestConfig constructor itself. 

So the intended use is that you can set it per request (with both openshift and in kubernetes) using code like this:

```
 NamespacedOpenShiftClient client;
 client.withRequestConfig(new RequestConfigBuilder(client.getConfiguration().getRequestConfig())
                .withImpersonateUsername("user")
                .build())
                .call(c -> c.namespaces().create(...));
```

**Comment by ibuziuk** [comment]
@lulf thanks for the input, but unfortunately  adding RequestConfig per request will not work for our use-case. Basically, we are building config per che workspace [1] and than using it for creating custom UnclosableOpenShiftClient [2]. What I think should help in our case is adding `ImpersonatorInterceptor` directly to our OkHttpClient [3] . 

[1] https://github.com/redhat-developer/rh-che/blob/master/plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/Fabric8WorkspaceEnvironmentProvider.java#L98
[2] https://github.com/eclipse/che/blob/7b4adc606e6a09721c45395405adc4689689e74c/infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftClientFactory.java#L191
[3] https://github.com/eclipse/che/blob/7b4adc606e6a09721c45395405adc4689689e74c/infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftClientFactory.java#L188

**Comment by lordofthejars** [comment]
@ibuziuk @lulf I have taken this issue to check if any development is required or not. I have not the deep details about this issue,  but after what both of you discussed It seems that the best way to proceed is that `ImpersonatorInterceptor` is added directly in che OkHttpClient instance. Is it feasible? Do you need anything from kubernetes client to do it? Or a different approach would be better?

**Comment by lulf** [comment]
@lordofthejars I think the fix is to add the impersonation parameters to the RequestConfig object, i.e impersonateUser, impersonateGroup, Map<String, String> impersonateExtra .

**Comment by lordofthejars** [comment]
But @ibuziuk says that it should not be by request

**Comment by lulf** [comment]
@lordofthejars I think there is a default RequestConfig, so its not per request. Making sure the impersonate settings  are copied in the same as other settings on the RequestConfig should be sufficient.

**Comment by lordofthejars** [comment]
@lulf 
>  I think the fix is to add the impersonation parameters to the RequestConfig object

`RequestConfig` already contains these parameters. What I have sen is that `Config` objet does not set them to the created `RequestConfig`, so the solution would be add them in `Config` constructor and set them?

**Comment by lulf** [comment]
@lordofthejars Yes, you are right, Config is where they should be added.

**Comment by lordofthejars** [comment]
@lulf @ibuziuk I have sent this PR https://github.com/fabric8io/kubernetes-client/pull/1117 to fix this problem. Can you review, please? I cannot add you as reviewer of the PR

**Comment by piyush1594** [comment]
Closed in #1117
